packages/core/src/interpreter.ts

```diff
@@ -129,160 +107,132 @@ export class Interpreter<
 
   // Actor Ref
   public _parent?: ActorRef<any>;
-  public name: string;
   public ref: ActorRef<TEvent>;
+  private _actorContext: ActorContext<TEvent, SnapshotFrom<TBehavior>>;
 
   /**
    * The globally unique process ID for this invocation.
    */
   public sessionId: string;
-  private forwardTo: Set<string> = new Set();
+
+  // TODO: remove
+  public _forwardTo: Set<AnyActorRef> = new Set();
 
   /**
-   * Creates a new Interpreter instance (i.e., service) for the given machine with the provided options, if any.
+   * Creates a new Interpreter instance (i.e., service) for the given behavior with the provided options, if any.
    *
-   * @param machine The machine to be interpreted
+   * @param behavior The behavior to be interpreted
    * @param options Interpreter options
    */
-  constructor(
-    public machine: StateMachine<
-      TContext,
-      TEvent,
-      any,
-      any,
-      TResolvedTypesMeta
-    >,
-    options?: InterpreterOptions
-  ) {
+  constructor(public behavior: TBehavior, options?: InterpreterOptions) {
     const resolvedOptions = {
       ...defaultOptions,
       ...options
-    } as Required<InterpreterOptions>;
+    };
 
     const { clock, logger, parent, id } = resolvedOptions;
+    const self = this;
 
-    const resolvedId = id !== undefined ? id : machine.key;
-
-    this.name = this.id = resolvedId;
+    // TODO: this should come from a "system"
+    this.sessionId = registry.bookId();
+    this.id = id ?? this.sessionId;
     this.logger = logger;
     this.clock = clock;
     this._parent = parent;
-
     this.options = resolvedOptions;
-
     this.ref = this;
-
-    this.sessionId = this.ref.name;
-  }
-
-  public get initialized() {
-    return this.status === InterpreterStatus.Running;
-  }
-
-  public get initialState(): State<TContext, TEvent, TResolvedTypesMeta> {
-    const initialState =
-      this._initialState ||
-      ((this._initialState = this.machine.getInitialState()),
-      this._initialState);
-    return initialState;
+    this._actorContext = {
+      self,
+      id: this.id,
+      sessionId: this.sessionId,
+      logger: this.logger,
+      exec: (fn) => {
+        if (self.status === InterpreterStatus.NotStarted) {
+          this._deferred.push(fn);
+        } else {
+          fn();
+        }
+      },
```

Comment:
This kind of a thing introduces a bug:
```ts
const m = createMachine({
    initial: 'a',
    states: {
    a: {
        entry: () => {
        console.log('entered a')
        },
        always: 'b'
    },
    b: {
    }
    }
});

const s = interpret(m)
s.getInitialState()
s.start('b')
```

The problem is that `s.getInitialState` "provides" the actor context to the inner `behavior.getInitialState`:
https://github.com/statelyai/xstate/blob/ec4a4c185bbb3ceccd8004d2930a83e0074e845e/packages/core/src/interpreter.ts#L166-L170
and that in turn "executes" actions immediately. The actions are not actually executed right away cause they are getting deferred here and later on executed within `update` here:
https://github.com/statelyai/xstate/blob/ec4a4c185bbb3ceccd8004d2930a83e0074e845e/packages/core/src/interpreter.ts#L177-L181

Reply:
Fixed here (with test): 42045534f6a04c5341cb9b65bb275a3a43b05104

Deferring actions is necessary in this current implementation (can refactor later), but the potential issue mentioned is fixed.